### PR TITLE
Better python printing

### PR DESCRIPTION
**Pull request by gchenfc**
Since `gtsam::KeyFormatter` is a `boost::function` and pybind11 does not do `boost::functions`, some modification is needed in order to be able to use the `print_(string s, KeyFormatter)` function.  Specifically, the `print_` function is altered so that the python interface accepts a `std::function` and then calls the gtsam `print` with the boost version.  So for example:

```c++
class ClassWithPrint {
  void print(const string &s,
             const gtsam::KeyFormatter &keyFormatter);
};
```
```c++
...
        .def("print_",
                    [](gtsam::ClassWithPrint* self, const string& s, const std::function<std::string(gtsam::Key)>& keyFormatter) {
                        py::scoped_ostream_redirect stream;
                        self->print(s, [&keyFormatter](gtsam::Key key){return keyFormatter(key);});
                    }, py::arg("s"), py::arg("keyFormatter"))
```

Note: `py::scoped_ostream_redirect stream;` is so that printing occurs to python sys.stdout which just generally plays nicer with python.  See [pybind docs](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/utilities.html#capturing-standard-output-from-ostream)

Also, this allows functionals for the keyformatter:
```python
def myKeyFormatter(key):
    return 'this is my key formatter {}'.format(key)
factor.print_('factor: ', myKeyFormatter)
```
outputs:
```
factor: min torque factor
  keys = { this is my key formatter 23924273508777984 }
  noise model: unit (1) 
```

### CODE REVIEW

File: cmake/GtwrapUtils.cmake [hunk 1]
```diff
@@ -103,3 +103,37 @@ macro(gtwrap_get_python_version)
...
+
+# Concatenate multiple wrapper interface headers into one.
+# The concatenation will be (re)performed if and only if any interface files
+# change.
+#
+# Arguments:
+# ~~~
+# destination: The concatenated master interface header file will be placed here.
+# inputs (optional): All the input interface header files
+function(combine_interface_headers
```
Comment (varunagrawal) [inline cmake/GtwrapUtils.cmake:115]: What's up with this? Not part of this PR.

**Review by varunagrawal — CHANGES_REQUESTED** [review]
Partial review. I am confused why not just add support for functionals?
https://pybind11.readthedocs.io/en/stable/advanced/cast/functional.html

Also, you've mixed up two PRs. Please separate, or wait till I add functional support, or even better, discuss?

File: cmake/GtwrapUtils.cmake (hunk 1 above)
Comment (gchenfc) [reply cmake/GtwrapUtils.cmake]: my bad... temporarily i've just changed the base of this PR, but if this goes through before the other PR then I'll make a new branch/PR to merge

File: gtwrap/pybind_wrapper.py [hunk 2]
```diff
@@ -126,6 +126,23 @@ def _wrap_method(self, method, cpp_class, prefix, suffix, method_suffix=""):
 
         if method.name == 'print':
             type_list = method.args.to_cpp(self.use_boost)
+            if len(type_list) == 2 \
+                    and 'string' in type_list[0] \
+                    and 'gtsam::KeyFormatter' in type_list[1]:
```
[the remark this answers is not in the document]
Comment (gchenfc) [reply gtwrap/pybind_wrapper.py:132]: not sure I understand your question - gtsam::KeyFormatter is typedef'd to boost::function, so it is a functional?  Unless that's not what a functional is...

**Comment by gchenfc** [comment]
Once https://github.com/borglab/gtsam/pull/735 merges, this will no longer be necessary so I'm closing this PR.
